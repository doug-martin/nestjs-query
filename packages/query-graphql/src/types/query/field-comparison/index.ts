export * from './boolean-field-comparison.type';
export * from './date-field-comparison.type';
export * from './float-field-comparison.type';
export * from './field-comparison.factory';
export * from './int-field-comparison.type';
export * from './number-field-comparison.type';
export * from './string-field-comparison.type';
export * from './timestamp-field-comparison.type';
