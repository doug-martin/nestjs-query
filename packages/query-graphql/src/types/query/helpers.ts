export const isInAllowedList = <T>(arr: T[] | undefined, val: T): boolean => arr?.includes(val) ?? true;
