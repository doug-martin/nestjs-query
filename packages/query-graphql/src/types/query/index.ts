export { QueryArgsType } from './query-args.type';
export {
  CursorQueryArgsType,
  OffsetQueryArgsType,
  NonePagingQueryArgsType,
  QueryArgsTypeOpts,
  NonePagingQueryArgsTypeOpts,
  OffsetQueryArgsTypeOpts,
  CursorQueryArgsTypeOpts,
  StaticQueryType,
  QueryType,
} from './query-args';
export {
  FilterType,
  DeleteFilterType,
  UpdateFilterType,
  SubscriptionFilterType,
  AggregateFilterType,
} from './filter.type';
export { CursorPagingType, OffsetPagingType, NonePagingType, PagingTypes, PagingStrategies } from './paging';
