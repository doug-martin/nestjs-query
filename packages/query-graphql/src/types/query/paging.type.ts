import { Class, Paging } from '@nestjs-query/core';
import { Min, Validate, IsPositive } from 'class-validator';
import { Field, InputType, Int } from '@nestjs/graphql';
import { cursorToOffset } from 'graphql-relay';
import { ConnectionCursorType, ConnectionCursorScalar } from '../cursor.scalar';
import { CannotUseWith, CannotUseWithout, IsUndefined } from '../validators';

export interface CursorPagingType extends Paging {
  before?: ConnectionCursorType;
  after?: ConnectionCursorType;
  first?: number;
  last?: number;
}

/** @internal */
let graphQLCursorPaging: Class<CursorPagingType> | null = null;

export const CursorPagingType = (): Class<CursorPagingType> => {
  if (graphQLCursorPaging) {
    return graphQLCursorPaging;
  }
  // based on https://github.com/MichalLytek/type-graphql/issues/142#issuecomment-433120114
  @InputType('CursorPaging')
  class GraphQLCursorPagingImpl implements CursorPagingType {
    @Field(() => ConnectionCursorScalar, {
      nullable: true,
      description: 'Paginate before opaque cursor',
    })
    @IsUndefined()
    @Validate(CannotUseWithout, ['last'])
    @Validate(CannotUseWith, ['after', 'first'])
    before?: ConnectionCursorType;

    @Field(() => ConnectionCursorScalar, {
      nullable: true,
      description: 'Paginate after opaque cursor',
    })
    @IsUndefined()
    @Validate(CannotUseWithout, ['first'])
    @Validate(CannotUseWith, ['before', 'last'])
    after?: ConnectionCursorType;

    @Field(() => Int, { nullable: true, description: 'Paginate first' })
    @IsUndefined()
    @IsPositive()
    @Min(1)
    @Validate(CannotUseWith, ['before', 'last'])
    first?: number;

    @Field(() => Int, { nullable: true, description: 'Paginate last' })
    @IsUndefined()
    // Required `before`. This is a weird corner case.
    // We'd have to invert the ordering of query to get the last few items then re-invert it when emitting the results.
    // We'll just ignore it for now.
    @Validate(CannotUseWithout, ['before'])
    @Validate(CannotUseWith, ['after', 'first'])
    @Min(1)
    @IsPositive()
    last?: number;

    get limit(): number | undefined {
      if (this.isForwardPaging) {
        return this.first || 0;
      }
      if (this.isBackwardPaging) {
        const { last = 0, before } = this;
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const offsetFromCursor = cursorToOffset(before!);
        const offset = offsetFromCursor - last;
        // Check to see if our before-page is underflowing past the 0th item
        if (offset < 0) {
          // Adjust the limit with the underflow value
          return Math.max(last + offset, 0);
        }
        return last;
      }
      return undefined;
    }

    get offset(): number | undefined {
      if (this.isForwardPaging) {
        const { after } = this;
        const limit = after ? cursorToOffset(after) + 1 : 0;
        return Math.max(limit, 0);
      }
      if (this.isBackwardPaging) {
        const { last, before } = this;
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const offset = last ? cursorToOffset(before!) - last : 0;

        // Check to see if our before-page is underflowing past the 0th item
        return Math.max(offset, 0);
      }
      return undefined;
    }

    private get isForwardPaging(): boolean {
      return !!this.first || !!this.after;
    }

    private get isBackwardPaging(): boolean {
      return !!this.last || !!this.before;
    }
  }
  graphQLCursorPaging = GraphQLCursorPagingImpl;
  return graphQLCursorPaging;
};
