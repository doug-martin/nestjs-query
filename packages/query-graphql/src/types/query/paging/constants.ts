export enum PagingStrategies {
  CURSOR = 'cursor',
  OFFSET = 'offset',
  NONE = 'none',
}
