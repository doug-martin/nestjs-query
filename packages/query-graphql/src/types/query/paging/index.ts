export { getOrCreateCursorPagingType } from './cursor-paging.type';
export { getOrCreateOffsetPagingType } from './offset-paging.type';
export { getOrCreateNonePagingType } from './none-paging.type';
export { PagingStrategies } from './constants';
export {
  OffsetPagingType,
  NonePagingType,
  CursorPagingType,
  PagingTypes,
  InferPagingTypeFromStrategy,
} from './interfaces';
