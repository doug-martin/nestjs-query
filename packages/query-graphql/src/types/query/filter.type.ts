import { Class, Filter, MapReflector } from '@nestjs-query/core';
import { InputType, Field } from '@nestjs/graphql';
import { Type } from 'class-transformer';
import { ValidateNested } from 'class-validator';
import { upperCaseFirst } from 'upper-case-first';
import { ResolverRelation } from '../../resolvers/relations';
import { createFilterComparisonType } from './field-comparison';
import { getDTONames, getGraphqlObjectName } from '../../common';
import { getFilterableFields, getQueryOptions, getRelations, SkipIf } from '../../decorators';
import { isInAllowedList } from './helpers';

const reflector = new MapReflector('nestjs-query:filter-type');

export type FilterTypeOptions = {
  allowedBooleanExpressions?: ('and' | 'or')[];
};
export type FilterableRelations = Record<string, Class<unknown>>;
export interface FilterConstructor<T> {
  hasRequiredFilters: boolean;
  new (): Filter<T>;
}

function getObjectTypeName<DTO>(DTOClass: Class<DTO>): string {
  return getGraphqlObjectName(DTOClass, 'No fields found to create FilterType.');
}

function getOrCreateFilterType<T>(
  TClass: Class<T>,
  name: string,
  filterableRelations: FilterableRelations = {},
): FilterConstructor<T> {
  return reflector.memoize(TClass, name, () => {
    const { allowedBooleanExpressions }: FilterTypeOptions = getQueryOptions(TClass) ?? {};
    const fields = getFilterableFields(TClass);
    if (!fields.length) {
      throw new Error(`No fields found to create GraphQLFilter for ${TClass.name}`);
    }
    const hasRequiredFilters = fields.some((f) => f.advancedOptions?.filterRequired === true);
    const isNotAllowedComparison = (val: 'and' | 'or') => !isInAllowedList(allowedBooleanExpressions, val);

    @InputType(name)
    class GraphQLFilter {
      static hasRequiredFilters: boolean = hasRequiredFilters;

      @ValidateNested()
      @SkipIf(() => isNotAllowedComparison('and'), Field(() => [GraphQLFilter], { nullable: true }))
      @Type(() => GraphQLFilter)
      and?: Filter<T>[];

      @ValidateNested()
      @SkipIf(() => isNotAllowedComparison('or'), Field(() => [GraphQLFilter], { nullable: true }))
      @Type(() => GraphQLFilter)
      or?: Filter<T>[];
    }

    const { baseName } = getDTONames(TClass);
    fields.forEach(({ propertyName, target, advancedOptions, returnTypeFunc }) => {
      const FC = createFilterComparisonType({
        FieldType: target,
        fieldName: `${baseName}${upperCaseFirst(propertyName)}`,
        allowedComparisons: advancedOptions?.allowedComparisons,
        returnTypeFunc,
      });
      const nullable = advancedOptions?.filterRequired !== true;
      ValidateNested()(GraphQLFilter.prototype, propertyName);
      Field(() => FC, { nullable })(GraphQLFilter.prototype, propertyName);
      Type(() => FC)(GraphQLFilter.prototype, propertyName);
    });
    Object.keys(filterableRelations).forEach((field) => {
      const FieldType = filterableRelations[field];
      if (FieldType) {
        const FC = getOrCreateFilterType(FieldType, `${name}${getObjectTypeName(FieldType)}Filter`);
        ValidateNested()(GraphQLFilter.prototype, field);
        Field(() => FC, { nullable: true })(GraphQLFilter.prototype, field);
        Type(() => FC)(GraphQLFilter.prototype, field);
      }
    });
    return GraphQLFilter as FilterConstructor<T>;
  });
}

function getFilterableRelations(relations: Record<string, ResolverRelation<unknown>>): FilterableRelations {
  const filterableRelations: FilterableRelations = {};
  Object.keys(relations).forEach((r) => {
    const opts = relations[r];
    if (opts && opts.allowFiltering) {
      filterableRelations[r] = opts.DTO;
    }
  });
  return filterableRelations;
}

export function FilterType<T>(TClass: Class<T>): FilterConstructor<T> {
  const { one = {}, many = {} } = getRelations(TClass);
  const filterableRelations: FilterableRelations = { ...getFilterableRelations(one), ...getFilterableRelations(many) };
  return getOrCreateFilterType(TClass, `${getObjectTypeName(TClass)}Filter`, filterableRelations);
}

export function DeleteFilterType<T>(TClass: Class<T>): FilterConstructor<T> {
  return getOrCreateFilterType(TClass, `${getObjectTypeName(TClass)}DeleteFilter`);
}

export function UpdateFilterType<T>(TClass: Class<T>): FilterConstructor<T> {
  return getOrCreateFilterType(TClass, `${getObjectTypeName(TClass)}UpdateFilter`);
}

export function SubscriptionFilterType<T>(TClass: Class<T>): FilterConstructor<T> {
  return getOrCreateFilterType(TClass, `${getObjectTypeName(TClass)}SubscriptionFilter`);
}

export function AggregateFilterType<T>(TClass: Class<T>): FilterConstructor<T> {
  return getOrCreateFilterType(TClass, `${getObjectTypeName(TClass)}AggregateFilter`);
}
