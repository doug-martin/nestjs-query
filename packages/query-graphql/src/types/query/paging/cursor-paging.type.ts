import { Min, Validate, IsPositive } from 'class-validator';
import { Field, InputType, Int } from '@nestjs/graphql';
import { Class } from '@nestjs-query/core';
import { ConnectionCursorType, ConnectionCursorScalar } from '../../cursor.scalar';
import { CannotUseWith, CannotUseWithout, IsUndefined } from '../../validators';
import { PagingStrategies } from './constants';
import { CursorPagingType } from './interfaces';

/** @internal */
let graphQLCursorPaging: Class<CursorPagingType> | null = null;
// eslint-disable-next-line @typescript-eslint/no-redeclare -- intentional
export const getOrCreateCursorPagingType = (): Class<CursorPagingType> => {
  if (graphQLCursorPaging) {
    return graphQLCursorPaging;
  }
  // based on https://github.com/MichalLytek/type-graphql/issues/142#issuecomment-433120114
  @InputType('CursorPaging')
  class GraphQLCursorPagingImpl implements CursorPagingType {
    static strategy: PagingStrategies.CURSOR = PagingStrategies.CURSOR;

    @Field(() => ConnectionCursorScalar, {
      nullable: true,
      description: 'Paginate before opaque cursor',
    })
    @IsUndefined()
    @Validate(CannotUseWithout, ['last'])
    @Validate(CannotUseWith, ['after', 'first'])
    before?: ConnectionCursorType;

    @Field(() => ConnectionCursorScalar, {
      nullable: true,
      description: 'Paginate after opaque cursor',
    })
    @IsUndefined()
    @Validate(CannotUseWithout, ['first'])
    @Validate(CannotUseWith, ['before', 'last'])
    after?: ConnectionCursorType;

    @Field(() => Int, { nullable: true, description: 'Paginate first' })
    @IsUndefined()
    @IsPositive()
    @Min(1)
    @Validate(CannotUseWith, ['before', 'last'])
    first?: number;

    @Field(() => Int, { nullable: true, description: 'Paginate last' })
    @IsUndefined()
    // Required `before`. This is a weird corner case.
    // We'd have to invert the ordering of query to get the last few items then re-invert it when emitting the results.
    // We'll just ignore it for now.
    @Validate(CannotUseWithout, ['before'])
    @Validate(CannotUseWith, ['after', 'first'])
    @Min(1)
    @IsPositive()
    last?: number;
  }
  graphQLCursorPaging = GraphQLCursorPagingImpl;
  return graphQLCursorPaging;
};
