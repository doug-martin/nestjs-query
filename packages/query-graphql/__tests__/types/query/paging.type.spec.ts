import { plainToClass } from 'class-transformer';
import { validateSync } from 'class-validator';
import { Resolver, Query, Int, InputType, Args } from '@nestjs/graphql';
import { generateSchema } from '../../__fixtures__';
import { getOrCreateCursorPagingType } from '../../../src/types/query/paging';

describe('PagingType', (): void => {
  const CursorPaging = getOrCreateCursorPagingType();
  it('should create the correct filter graphql schema', async () => {
    @InputType()
    class Paging extends CursorPaging {}

    @Resolver()
    class PagingTypeSpec {
      @Query(() => Int)
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      test(@Args('input') input: Paging): number {
        return 1;
      }
    }
    const schema = await generateSchema([PagingTypeSpec]);
    expect(schema).toMatchSnapshot();
  });

  it('throw a validation error if first is defined with before', () => {
    const paging = plainToClass(CursorPaging, {
      first: 10,
      before: 'YXJyYXljb25uZWN0aW9uOjEx',
    });
    expect(validateSync(paging)).toEqual([
      {
        children: [],
        constraints: {
          CannotUseWith: 'Cannot be used with `after` , `first`.',
          CannotUseWithout: 'Cannot be used without `last`.',
        },
        property: 'before',
        target: {
          before: 'YXJyYXljb25uZWN0aW9uOjEx',
          first: 10,
        },
        value: 'YXJyYXljb25uZWN0aW9uOjEx',
      },
      {
        children: [],
        constraints: {
          CannotUseWith: 'Cannot be used with `before` , `last`.',
        },
        property: 'first',
        target: {
          before: 'YXJyYXljb25uZWN0aW9uOjEx',
          first: 10,
        },
        value: 10,
      },
    ]);
  });

  it('throw a validation error if last is defined with after', () => {
    const paging = plainToClass(CursorPaging, {
      last: 10,
      after: 'YXJyYXljb25uZWN0aW9uOjEx',
    });
    expect(validateSync(paging)).toEqual([
      {
        children: [],
        constraints: {
          CannotUseWith: 'Cannot be used with `before` , `last`.',
          CannotUseWithout: 'Cannot be used without `first`.',
        },
        property: 'after',
        target: {
          after: 'YXJyYXljb25uZWN0aW9uOjEx',
          last: 10,
        },
        value: 'YXJyYXljb25uZWN0aW9uOjEx',
      },
      {
        children: [],
        constraints: {
          CannotUseWith: 'Cannot be used with `after` , `first`.',
          CannotUseWithout: 'Cannot be used without `before`.',
        },
        property: 'last',
        target: {
          after: 'YXJyYXljb25uZWN0aW9uOjEx',
          last: 10,
        },
        value: 10,
      },
    ]);
  });

  it('throw a validation error if after is defined without first', () => {
    const paging = plainToClass(CursorPaging, {
      after: 'YXJyYXljb25uZWN0aW9uOjEx',
    });
    const validateErrors = validateSync(paging);
    expect(validateErrors).toEqual([
      {
        children: [],
        constraints: {
          CannotUseWithout: 'Cannot be used without `first`.',
        },
        property: 'after',
        target: {
          after: 'YXJyYXljb25uZWN0aW9uOjEx',
        },
        value: 'YXJyYXljb25uZWN0aW9uOjEx',
      },
    ]);
  });

  it('throw a validation before if after is defined without last', () => {
    const paging = plainToClass(CursorPaging, {
      before: 'YXJyYXljb25uZWN0aW9uOjEx',
    });
    const validateErrors = validateSync(paging);
    expect(validateErrors).toEqual([
      {
        children: [],
        constraints: {
          CannotUseWithout: 'Cannot be used without `last`.',
        },
        property: 'before',
        target: {
          before: 'YXJyYXljb25uZWN0aW9uOjEx',
        },
        value: 'YXJyYXljb25uZWN0aW9uOjEx',
      },
    ]);
  });
});
