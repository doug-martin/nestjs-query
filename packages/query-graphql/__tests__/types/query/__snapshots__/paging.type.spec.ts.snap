// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PagingType should create the correct filter graphql schema 1`] = `
type Query {
  test(input: Paging!): Int!
}

input Paging {
  """Paginate before opaque cursor"""
  before: ConnectionCursor

  """Paginate after opaque cursor"""
  after: ConnectionCursor

  """Paginate first"""
  first: Int

  """Paginate last"""
  last: Int
}

"""Cursor for paging through collections"""
scalar ConnectionCursor

`;
